### PR TITLE
Add a way to manage API keys

### PR DESCRIPTION
**Pull request by Karakatiza666**
Is this a user-visible change (yes/no): yes
Fix https://github.com/feldera/feldera/issues/1103

image
image

### CODE REVIEW

File: crates/pipeline_manager/src/api/auth_api.rs [hunk 1]
```diff
@@ -8,7 +8,7 @@ use super::{ManagerError, ServerState};
 
 /// Get authentication provider configuration
 #[utoipa::path(
-    path="/config/authentication",
+    path="/../config/authentication",
```
Comment (lalithsuresh) [inline crates/pipeline_manager/src/api/auth_api.rs:11]: Why is this change required? This was fixed to `/config/authentication` as part of https://github.com/feldera/feldera/issues/1084.

File: web-console/src/lib/components/settings/CreateApiKeyDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+          onSuccess={data => generateApiKey(data)}
+          FormProps={{ 'data-testid': 'box-form-generate-api-key' } as any}
+        >
+          <Stack spacing={4}>
+            <DialogContentText color={theme.palette.warning.contrastText}>
+              You will only see API key secret once after generation.
+              <br />
+              You will not be able to view API key secret afterwards.
+              <br />
+              You should store generated API key yourself.
```
Comment (lalithsuresh) [inline web-console/src/lib/components/settings/CreateApiKeyDialog.tsx:85]:
```
You will be shown the generated API key only once. You will not be able view it afterwards. Please store it securely. 
```

**Review by lalithsuresh — COMMENTED** [review]
Regarding the side bar. I don't think "ecosystem" is the right section for these settings.

API key generation should only be exposed by the UI when using authentication. It should be part of the user profile drop down on the top right.

File: crates/pipeline_manager/src/api/auth_api.rs (hunk 1 above)
Comment (Karakatiza666) [reply crates/pipeline_manager/src/api/auth_api.rs:11]: Otherwise TypeScript generates a url /v0/config/authentication for this action, which is wrong. I understand this path is relative (a subpath), hence the issue

**Comment by Karakatiza666** [comment]
Would it make sense to leave API keys view on the settings page and only show it when authenticated?
I am wary of moving the entire list to account dropdown.

**Comment by Karakatiza666** [comment]
image

We will need settings page at some point anyway

File: crates/pipeline_manager/src/api/auth_api.rs (hunk 1 above)
Comment (lalithsuresh) [reply crates/pipeline_manager/src/api/auth_api.rs:11]: Ah, now I see why you need this.

Our use of a `Servers` variable is the issue: https://github.com/feldera/feldera/blob/main/crates/pipeline_manager/src/api/mod.rs#L63

I'll see if I can find a quick workaround so you don't need this change. Basically, we need the `/v0` prefix only for the authenticated APIs.

**Comment by lalithsuresh** [comment]
> Would it make sense to leave API keys view on the settings page and only show it when authenticated? I am wary of moving the entire list to account dropdown.

I'm fine with that too.

File: crates/pipeline_manager/src/api/auth_api.rs (hunk 1 above)
Comment (lalithsuresh) [reply crates/pipeline_manager/src/api/auth_api.rs:11]: @Karakatiza666 btw, I've been using the authentication code from the UI just fine. What workflow breaks without the relative pathing?

Comment (Karakatiza666) [reply crates/pipeline_manager/src/api/auth_api.rs:11]: You are using non-updated UI code. I re-generated UI OpenAPI bindings and there this issue popped up, so I had to add '/../' to get generated code to behave as before.
I think I added this previously too, actually.

Comment (lalithsuresh) [reply crates/pipeline_manager/src/api/auth_api.rs:11]: Ack. I've opened https://github.com/feldera/feldera/pull/1139. See if that fixes your problem. If so, let's merge that and rebase this PR on top of it.

**Review by lalithsuresh — APPROVED** [review]
LGTM from my side.

**Comment by gz** [comment]
this goes to an empty page for me? is it because I have no auth?

image

> API key generation should only be exposed by the UI when using authentication. It should be part of the user profile drop down on the top right.

I felt this would be a good place to put this (fwiw in github it's also under user)

File: web-console/src/lib/components/settings/ApiKeyList.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      </Button>
+    </Box>
+  )
+  const title = <Typography variant='h6'>API keys</Typography>
+  if (isPending) {
+    return (
+      <Stack spacing={4}>
+        {title}
+        <Box sx={{ display: 'flex', alignItems: 'center', gap: 2 }}>
+          <Icon270RingWithBg></Icon270RingWithBg>fetching keys...
```
Comment (gz) [inline web-console/src/lib/components/settings/ApiKeyList.tsx:40]:
```suggestion
          <Icon270RingWithBg></Icon270RingWithBg>Fetching keys...
```

File: web-console/src/lib/components/settings/ApiKeyList.tsx [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        </Box>
+      </Stack>
+    )
+  }
+  if (!apiKeys?.length) {
+    return (
+      <Stack spacing={4}>
+        {title}
+        No api keys generated
```
Comment (gz) [inline web-console/src/lib/components/settings/ApiKeyList.tsx:49]:
```suggestion
        No API keys generated
```

**Review by gz — APPROVED** [review]

**Comment by gz** [comment]
I can't see it right now in the browser, but judging from the screenshots I don't think there is need to have a scrollbar here?

image

**Comment by Karakatiza666** [comment]
The scrollbar only appears to scroll through more than 4 keys

**Comment by Karakatiza666** [comment]
@gz yes, the keys should only be accessible when auth is enabled. Authentication is the only section in settings, but only temporarily. There is no point to add code to hide the page when it will get filled with content relatively soon
